001-software-supply-chain-integrity.md

```diff
@@ -0,0 +1,169 @@
+value are placing themselves at substantial personal risk.
+
+To address this we must avoid trusting any single human or any system they
+control by design, in order to protect those individuals and downstream users.
+
+## Assumptions
+
+* Any one human or computer involved in the supply chain is compromised
+* All systems managed by a single party (IT, third parties) are compromised
+* Any code or binaries controlled by one system or party are compromised
+* All memory of all internet connected computers is visible to adversary
+* Any logging or backups that are not signed are tampered with
+* Adversary wields a "zero-click" "Zero-Day" exploit for any system
+
+## Requirements
+
+The following only applies if code is bound for production, and these can be
+met in any order or cadence desired.
+
+* Third party code:
+  * MUST have extensive and frequent review.
+    * Example: The Linux kernel has well funded distributed review.
+  * MUST be hash-pinned at known reviewed versions
+  * MUST be at version with all known related security patches
+  * SHOULD be latest versions if security disclosures lag behind releases
+    * Example: The Linux kernel
+* First party code:
+  * MUST be signed in version control systems by well known author keys
+  * MUST be signed by separate subject matter expert after security review
+* All code MUST build deterministically
+* All binaries:
+  * MUST be built and signed by multiple parties with no management overlay
+    * Example: One build by IT, another by Infrastructure team managed CI/CD
+  * MUST be signed by well known keys signed by a common CA
+    * Example: PGP Smartcards signed under OpenPGP-CA.
+* All signing keys:
+  * MUST be stored in approved PERSONAL HSMs
+  * MUST NOT ever come in contact with network accessible memory
+* All execution environments SHOULD be able to verify m-of-n binary signatures
+  * Example: Custom Secure Boot verifies minimum signatures against CA
+* Only code required for sensitive operation SHOULD be present
+  * Example: A server has no need for sound card drivers
+
+## Design
+
+Many implementations are possible under the above requirements and assumptions
+however the following opinionated workflow can be used as a reference to
+consider.
+
+### Developer Workflow
+
+1. Software Engineer makes source code change
+    * Changes that bump third party dependencies should make them easy to review
+      * Link to Changelog indicating this is the latest appropriate release
+      * Links to evidence of third party review
+      * Link to source of truth for hashes or keys as appropriate
+      * Hash of commit or archive that is verified at build time
+      * Public key hashes of any keys that can be used to verify sources
+    * Automated verification of hash and signature in build logic
+2. Software Engineer builds and tests changes locally
+    * Local build should put output hashes of generated binaries for committing
```

Comment:
this might be fun given that Talos embeds git-based version into the binary, and final version (after merge) is not known at build time on a branch